### PR TITLE
[Merged by Bors] - Subscribe to altair gossip topics 2 slots before fork

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

N/A

## Proposed Changes

Add a fork_digest to `ForkContext` only if it is set in the config.
Reject gossip messages on post fork topics before the fork happens.

Edit: Instead of rejecting gossip messages on post fork topics, we now subscribe to post fork topics 2 slots before the fork.

### CODE REVIEW

**Comment by pawanjay176** [comment]
@michaelsproul can we tweak the `ChainSpec` deserialisation such that if `ALTAIR_FORK_EPOCH == u64::max()` in config.yaml, we set `chain_spec.altair_fork_epoch = None`. Didn't see any issue with that after a brief look.

**Comment by michaelsproul** [comment]
> @michaelsproul can we tweak the `ChainSpec` deserialisation such that if `ALTAIR_FORK_EPOCH == u64::max()` in config.yaml, we set `chain_spec.altair_fork_epoch = None`. Didn't see any issue with that after a brief look.

Yeah I think this would fine. No fork is _actually_ going to activate at `u64::max()`, so it makes sense to map that to `None`. It might be nice to get this changed in the spec though. They could set it to `null` in the YAML.

**Review by AgeManning — COMMENTED** [review]
There's a few things we probably need to consider here.

If we receive messages on a topic we are subscribed and then reject them, this is considered censoring and the gossipsub scoring algo will penalize us. The correct approach is to only subscribe to a topic when we want to receive messages. 

If there is a period of time we dont want to receive messages, we should not subscribe to the topic. In this case it seems we dont want to subscribe to the topic until the fork.

If we do want to keep this current logic, it might also be nicer to filter the messages at the behaviour level. The behaviour has a `fork_context` so I think its possible to filter the messages as we get them, rather than sending them up to the `network` to do it for us.

File: consensus/types/src/chain_spec.rs [hunk 1]
```diff
@@ -582,6 +584,20 @@ impl Default for Config {
...
 }
 
+/// Util function to deserialize a u64::max() fork epoch as `None`.
+fn deserialize_fork_epoch<'de, D>(deserializer: D) -> Result<Option<MaybeQuoted<Epoch>>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let decoded: Option<MaybeQuoted<Epoch>> = serde::de::Deserialize::deserialize(deserializer)?;
+    if let Some(fork_epoch) = decoded {
+        if fork_epoch.value != Epoch::max_value() {
```
Comment (michaelsproul) [inline consensus/types/src/chain_spec.rs:610]: Unfortunately we can't do this hack in a backwards-compatible way. If you run a Lighthouse v1.5.0 VC on a network without Altair (e.g. Prater/mainnet) against a BN running this branch then it fails because of the spec discrepancy:

```
Aug 26 18:44:49.872 INFO Connected to beacon node                endpoint: http://localhost:5052/, version: Lighthouse/v1.5.0-00bdafe/x86_64-linux
Aug 26 18:44:49.872 ERRO The beacon node is using a different Eth2 specification to this validator client. See the --network command., endpoint: http://localhost:5052/
```

It might be simpler to revert https://github.com/sigp/lighthouse/pull/2532/commits/00bdafeb167c880a9201044f453b2b90217ccaad, or if there are other changes required, add a method to `ChainSpec` that returns the "effective altair fork epoch" like:

```rust
pub fn effective_altair_fork_epoch(&self) -> Option<Epoch> {
    self.altair_fork_epoch.copied().filter(|epoch| epoch != Epoch::max_value())
}
```

**Review by michaelsproul — CHANGES_REQUESTED** [review]

**Comment by pawanjay176** [comment]
> If there is a period of time we dont want to receive messages, we should not subscribe to the topic. In this case it seems we dont want to subscribe to the topic until the fork.

I subscribed to all topics before the fork mainly because the spec [says](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/p2p-interface.md#transitioning-the-gossip) 
```
The topics that are not removed in a fork are updated with a new ForkDigestValue. In advance of the fork, a node SHOULD subscribe to the post-fork variants of the topics.

Subscriptions are expected to be well-received, all updated nodes should subscribe as well. Topic-meshes can be grafted quickly as the nodes are already connected and exchanging gossip control messages.
```

We could also interpret it as subscribe to the post-fork topics a few epochs in advance. But we would still have to deal with the same problem of rejecting messages only for a shorter duration.


> If we receive messages on a topic we are subscribed and then reject them, this is considered censoring and the gossipsub scoring algo will penalize us. The correct approach is to only subscribe to a topic when we want to receive messages.

I think this is fine because a node sending us post fork messages before the fork slot is either malicious or has clock issues. So we probably wouldn't want to peer with such a node anyway.

> If we do want to keep this current logic, it might also be nicer to filter the messages at the behaviour level. 

I started out implementing this but it got ugly because the behaviour doesn't have the slot clock which we need to check if we are within `MAXIMUM_GOSSIP_CLOCK_DISPARITY` of the fork instant. 

We could simplify things by not accounting for `MAXIMUM_GOSSIP_CLOCK_DISPARITY` and just checking if the topic on which the message was received is the one corresponding to the current fork. But this would again cause issues because we will end up downscoring peers sending messages on pre-fork topics after the fork. This is possible due to lagging IWANT messages as mentioned in the spec.
```
Peers who propagate messages on the pre-fork topics MUST NOT be scored negatively. Lagging IWANT may force them to.
```
With all this, I felt it was the cleanest to handle it in the network service. Please let me know if there's a better way of handling it :pray:

File: consensus/types/src/chain_spec.rs (hunk 1 above)
Comment (pawanjay176) [reply consensus/types/src/chain_spec.rs:610]: Doing the filtering in the `ForkContext` constructor like before now.

Comment (michaelsproul) [reply consensus/types/src/chain_spec.rs:610]: Ah! I think I've thought of how to do it in a backwards-compatible way. Because the specs always use `u64::max` to denote our `None`, we can make `None` the Lighthouse internal representation for `u64::max`. I.e. we keep your `Deserialize` implementation and add a `Serialize` implementation for `altair_fork_epoch` that converts `None` to `u64::max`. We could make the `altair_fork_epoch` a regular `Epoch` on the `Config` struct.

In other words: bijection between `Option<Epoch>` and `Epoch` where `None <-> Epoch::max_value()` and `Some(i) <-> i if i < Epoch::max_value()`

**Review by michaelsproul — APPROVED** [review]
The fork epoch stuff is looking good :ok_hand:

**Comment by AgeManning** [comment]
I think its still not right to subscribe to a topic and then try and downscore peers who send us information on a topic we've subscribed to. Although this may work at a lighthouse level, we're breaking a few things in gossipsub, which is designed to give you messages for which you are subscribed (i.e we will also get gossip (IHAVE messages) from random other peers subscribed to the topic if they accept them at the gossip level, and i'm worried we could be banning misc peers). 

If we don't want messages from a topic, we shouldn't subscribe to it. Building the mesh should be relatively quick (so we don't need much time to subscribe in advance). From reading the spec, it just says we should subscribe in advance (which I agree with) but maybe like 1 or two slots in advance of the fork. This way we don't have to worry about which topic we get messages on, we let gossipsub filter things we don't want out, simply by our subscriptions (all the gossipsub scoring logic should remain fine this way also). 

In terms of implementation, I've not looked closely at the gossipsub subscription timing logic. But I imagine the `Network` service keeps track of when we are forking, and potentially it can have an event that fires 1/2 slots before the fork to tell the network to subscribe to the new topics. Then that's all that needs to be done and gossipsub will filter everything for us before this time. 

I could be missing something, so let me know if there's other things you think we should consider here.

**Comment by pawanjay176** [comment]
@AgeManning Agree with your points above. Reverted the gossipsub filtering and added a new timer that fires `SUBSCRIBE_DELAY_SLOTS = 2` slots before the fork. We then subscribe to all currently subscribed topics with the new fork_digest.

We subscribe to both base and altair fork_digests if the beacon node is started between `fork_slot - SUBSCRIBE_DELAY_SLOTS` and `fork_slot` and just altair fork_digest  after the fork.

Tested all scenarios on local testnets and the simulator and it is working as expected.

**Comment by paulhauner** [comment]
I ran this for the Prater Altair fork and it went well!

Just for clarity, my understanding is that we're blocked on approval from @AgeManning. No rush, just wanted to make it clear.

**Review by AgeManning — APPROVED** [review]
We discussed this internally and decided we should merge this for mainnet and make a side PR to shift some of the logic into the behaviour.

**Comment by AgeManning** [comment]
@pawanjay176 - Could you merge in the latest unstable and resolve the conflicts, then I think we're good to go.

**Review by gagadbug — CHANGES_REQUESTED** [review]
- [ ] kdkslL
